### PR TITLE
Fix #640: Implement C-like escape sequences for c"..." literals

### LINKED ISSUE #640: cstring constant are not escaped
Scala-Native (0.1.0)

The first example produce an invalid llvm output

```
[error] /home/gui/hands-on-scala-native/target/scala-2.11/ll/__const.ll:763:48: error: constant expression type mismatch
[error] @"__const:: 353358909.10" = constant [17 x i8] c"%s \\t %.2f %s/s\00"
[error]                                                ^
[error] 1 error generated.

...

[error] clang-3.9: error: no such file or directory: '/home/gui/hands-on-scala-native/target/scala-2.11/ll/__const.ll.o'
```

```scala
def mvwprintw(window: Ptr[Window], y: CInt, x: CInt, fmt: CString, args: CVararg*): CInt = extern

// fails
mvwprintw(
  window,
  line,
  1,
  c"%s \\t %.2f %s/s",
  toCString(label),
  rate,
  toCString(unit)
)

// pass
mvwprintw(
  window,
  line,
  1,
  c"foo",
  toCString(label),
  rate,
  toCString(unit)
)

// pass
mvwprintw(
  window,
  line,
  1,
  toCString("%s \\t %.2f %s/s"),
  toCString(label),
  rate,
  toCString(unit)
)
```

### PR DESCRIPTION
**Pull request by matil019**
Fix: #640 

This PR contains questions. Please see `TODO` in the changed file.

This PR includes two changes:
* Count bytes of CString constants when they contain escape sequences
* Format CString constants into proper LLVM IR string constants  
  e.g. `\t` -> `\09`

### CODE REVIEW

File: nir/src/main/scala/scala/scalanative/nir/Vals.scala [hunk 1]
```diff
@@ -28,6 +28,47 @@ sealed abstract class Val {
     case Val.String(_) => Rt.String
   }
 
+  private def countBytes(s: String): Int = {
+    import Character.isDigit
+
+    // TODO: should be a compilation error, how to report?
+    def malformed() = ???
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Vals.scala]: `throw new IllegalArgumentException("malforrmed C string: " + string)` is the way to crash here.

File: nir/src/main/scala/scala/scalanative/nir/Vals.scala [hunk 2]
```diff
@@ -28,6 +28,47 @@ sealed abstract class Val {
...
+
+    // Subtracts from the length of the bytes for each escape sequence
+    // uses String, not Seq[Byte], but should be okay since we handle ASCII only
+    def uncountEscapes(from: Int, accum: Int): Int =
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Vals.scala]: Needs `@tailrec` annotation.

File: tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala [hunk 3]
```diff
@@ -394,6 +392,56 @@ object CodeGen {
         unsupported(v)
     }
 
+    def genChars(value: String): Unit = {
+      // `value` should contain a content of a CString literal as is in its source file
+      // malformed literals are assumed absent
+      str("c\"")
+      def loop(from: Int): Unit =
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala]: `@tailrec`.

**Comment by matil019** [comment]
Where do `nir` and/or `tools` go into? `tools/src/test/scala/scala/scalanative/FrameworkTest.scala` or similar classes?

**Comment by densh** [comment]
@matil019 `unit-tests` tests should work for this, as it wouldn't compile if something went wrong.

**Comment by matil019** [comment]
@densh Can't get `nir` and `tools` to go into the classpath for `unit-tests`.

**Comment by densh** [comment]
@matil019 The idea would be to just have a few tests that use c strings with escape sequences in the test source code. You do not need nir and/or tools to be on the classpath.

**Comment by matil019** [comment]
@densh Ah! I didn't have that idea. Thanks, I'll try that.

**Comment by matil019** [comment]
One point: `fromCString(c"\"")` is rejected by the Scala compiler (before compiled into LLVM) for some reason. Seeing that `s"\""` is rejected as well, I suspect it comes from the restriction of Scala's String Interpolation.

**Comment by densh** [comment]
LGTM, thanks @matil019 ! 👍

**Pull request merged**